### ISSUE TITLE
replacing TunnelAdapter

### ISSUE THREAD

**Issue by happyleavesaoc** [construct/construct#404]
I'm upgrading a parser from `2.5` to `2.8`. I have a `TunnelAdapter` used like so:

```py
header = "header"/Struct(
    "header_length"/Int32ul,
    "chapter_address"/Int32ul,
    Embedded(TunnelAdapter(
        "header"/String(lambda ctx: ctx._.header_length - 8, encoding = ZlibCompressor()),
        "header"/Struct(
            # parse header fields ...,
            Terminated
            )
        ))
    )
```

I know I need to use `Compressed`, but I'm not sure how to specify the length. I thought to use `Prefixed` as noted elsewhere, but I have another (non-compressed) field between the `header_length` and the compressed header.

What approach should I take?

**Comment by arekbulski** [comment]
I would suggest using Compressed with Bytes(this.headerlength-8) or something like that.
http://construct.readthedocs.io/en/latest/tunneling.html#compression-and-checksuming

**Comment by arekbulski** [comment]
Your code is very convoluted and hard to read, so I cannot help much. I would warn against Embedding anything, much less a Tunnel tho.

**Comment by happyleavesaoc** [comment]
`Compressed` consumes the whole stream unless used with `Prefixed`. After this header, I have more bytes, so I can't have `Compressed` read the remainder of the stream. Here's (hopefully!) as simpler example:

```python
my_parser = "header"/Struct(
    "header_length"/Int32ul,
    "chapter_address"/Int32ul,
    "decompressed_bytes"/Compressed(Bytes(this.header_length - 8), "zlib"),
    "another_field"/Int32ul
)
```
In this example, the `zlib` decompression fails because it consumes the bytes meant for `another_field` as well. I also can't used `Prefixed`, because my length field does not immediately prefix the compressed bytes.

**Comment by happyleavesaoc** [comment]
Just to follow up, I ended up making my own `Subconstruct` just like `Prefixed`, except the first argument is a length function instead of a length field. That way I could pass in any value instead of relying on the prefixed bytes.

One other note - would be cool if `Compressed` accepted a custom Compressor like `TunnelAdapter` did. My particular compressed bytes were zlib, but without the 2 zlib header bytes,   so I made my own `Compressed` version, modeled after the existing one.
